### PR TITLE
BUG: test all ufunc.types for return type, fix for exp, log

### LINKED ISSUE #13493: BUG: exp and log cast float16 to float32
After #13134 is merged, `np.exp` and `np.log` return float32 arrays for float16 inputs.  Is the change intended?  I could not find docs for it (https://www.numpy.org/devdocs/reference/generated/numpy.exp.html), while it is documented that `np.sum` casts ints to the default platform integer.

### Reproducing code example:

```python
import numpy as np
print(np.exp(np.array([1, 2], np.float16)).dtype)  # => float32
print(np.log(np.array([1, 2], np.float16)).dtype)  # => float32
```

### Numpy/Python version information:

```
1.17.0.dev0+64b3e67 3.7.1 (default, Nov  6 2018, 21:02:07)
[Clang 10.0.0 (clang-1000.10.44.4)]
```

### PR DESCRIPTION
**Pull request by mattip**
Fixes #13493 by registering a non-simd loop for 'e' (half) before the 'f' (float) one.

Also added a test that all ufuncs respect the types they declare

### CODE REVIEW

File: numpy/core/tests/test_ufunc.py [hunk 1]
```diff
@@ -1936,3 +1936,38 @@ def test_nat_is_not_inf(self, nat):
...
+            res = ufunc(*args)
+        if isinstance(res, tuple):
+            if ',' in outt:
+                outs = [t.strip() for t in outt.split(',')]
+            else:
+                outs = [t for t in outt]
+                assert len(res) == len(outs)
+                for r, t in zip(res, outs):
+                    assert r.dtype == np.dtype(t)
+            else:
```
Comment (charris) [inline numpy/core/tests/test_ufunc.py]: Indentation bug.

Comment (mattip) [reply numpy/core/tests/test_ufunc.py]: fixed

File: numpy/core/tests/test_ufunc.py [hunk 2]
```diff
@@ -1936,3 +1936,38 @@ def test_nat_is_not_inf(self, nat):
...
+    for them.
+
+    Choose the shape so even dot and matmul will succeed
+    '''
+    for typ in ufunc.types:
+        # types is a list of strings like ii->i
+        if 'O' in typ or '?' in typ:
+            continue
+        inp, out = typ.split('->')
+        if ',' in inp:
```
Comment (charris) [inline numpy/core/tests/test_ufunc.py]: I don't see any arguments like that, what does it work for?

File: numpy/core/tests/test_ufunc.py [hunk 3]
```diff
@@ -1936,3 +1936,38 @@ def test_nat_is_not_inf(self, nat):
...
+            args = [np.zeros((3, 3), t.strip()) for t in inp.split(',')]
+        else:
+            args = [np.ones((3, 3), t.strip()) for t in inp]
+        outt = out.strip()
```
Comment (charris) [inline numpy/core/tests/test_ufunc.py]: What if there are multiple outputs, possibly of different types? There are 23 like that.

File: numpy/core/tests/test_ufunc.py [hunk 4]
```diff
@@ -1936,3 +1936,38 @@ def test_nat_is_not_inf(self, nat):
...
+        with warnings.catch_warnings(record=True):
+            warnings.filterwarnings("always")
+            res = ufunc(*args)
+        if isinstance(res, tuple):
+            if ',' in outt:
```
Comment (charris) [inline numpy/core/tests/test_ufunc.py]: I don't see any with ','.

File: numpy/core/tests/test_ufunc.py (hunk 3 above)
Comment (charris) [reply numpy/core/tests/test_ufunc.py]: I see this is handled below.

**Comment by charris** [comment]
I don't see any types with ',', or that need to be stripped.

**Comment by mattip** [comment]
> I don't see any types with ',', or that need to be stripped.

It seems I got confused with gufunc signature, which has richer syntax than type signature. The latter contains only type letters and `->`. Fixing

File: numpy/core/tests/test_ufunc.py [hunk 5]
```diff
@@ -1936,3 +1936,31 @@ def test_nat_is_not_inf(self, nat):
...
+
+    Choose the shape so even dot and matmul will succeed
+    '''
+    for typ in ufunc.types:
+        # types is a list of strings like ii->i
+        if 'O' in typ or '?' in typ:
+            continue
+        inp, out = typ.split('->')
+        args = [np.ones((3, 3), t) for t in inp]
+        with warnings.catch_warnings(record=True):
```
Comment (charris) [inline numpy/core/tests/test_ufunc.py]: What happens to all the recorded warnings?

Comment (charris) [reply numpy/core/tests/test_ufunc.py]: Could get the same effect by just ignoring them :)

Comment (mattip) [reply numpy/core/tests/test_ufunc.py]: made the ignore explicit in the next line. Without `record=True` the test passes but all the warnings are printed out

Comment (charris) [reply numpy/core/tests/test_ufunc.py]: Shouldn't be printed out if they are ignored.

Comment (mattip) [reply numpy/core/tests/test_ufunc.py]: hmm, filterwarnings('ignore') is not supported by pytest. @charris what change are you suggesting?

Comment (charris) [reply numpy/core/tests/test_ufunc.py]: Need to use `suppress_warnings`. @seberg Do you know if we still need that for Python >= 3.5?

Comment (charris) [reply numpy/core/tests/test_ufunc.py]: The failing test is one of ours that looks for "ignore" because using that used to mess up the warnings module.

Comment (mattip) [reply numpy/core/tests/test_ufunc.py]: The code in this PR follows the pattern in [other places in the tests](https://github.com/numpy/numpy/blob/v1.16.3/numpy/core/tests/test_multiarray.py#L6980). Maybe we can accept this PR as-is and explore best practices for warnings in a separate issue/PR

Comment (seberg) [reply numpy/core/tests/test_ufunc.py]: Catch warnings is OK, it just catches all warnings, which is not ideal (and `suppress_warnings` can be nested and works reasonable then). Should double check, but the doc I put on `suppress_warnings` indeed suggests that `"ignore"` filters should be fine in numpy 1.17. I remember there were still some oddities, but I do not think they were relevant for this.

Comment (charris) [reply numpy/core/tests/test_ufunc.py]: @seberg Sounds like we can disable the test and clean things up after we drop 1.16 support.

**Comment by charris** [comment]
Thanks Matti.

File: numpy/core/tests/test_ufunc.py (hunk 5 above)
Comment (seberg) [reply numpy/core/tests/test_ufunc.py]: @charris do you mean the "no ignore warnings test"? This here should work in any case, maybe not the prettiest, but `suppress_warnings` is also not perfect.

**Pull request merged**